File: Leetcode/Longest Valid Parentheses/Longest_Valid_Parentheses.java

// Java program to find the length of longest valid parentheses
import java.io.*;
import java.util.*;
import java.math.*;

class Solution {
    public static int longestValidParentheses(String s) {
        int n = s.length();
 
       
        Stack<Integer> st = new Stack<>();

        //Initializing the stack to -1
        st.push(-1);
 
        
        int res = 0;
 
        //for loop to extract each character and check
        for (int i = 0; i < n; i++) 
        {
           
            if (s.charAt(i) == '(')
            //pusing into stack
                st.push(i);
 
          
            else
            {
                //poping out from stack
                    st.pop();
 
               //if stack is not empty then max is stored in res
                if (!st.empty())
                    res = Math.max(res, 
                                   i - st.peek());
 
               
                else
                    st.push(i);
            }
        }
        // return the result
        return res;
}

public static void main(String args[]) {
    Scanner sc = new Scanner(System.in);
    System.out.println("Enter parentheses");
    String s = sc.nextLine();
    System.out.println(longestValidParentheses(s));
}
}

File: Leetcode/Longest Valid Parentheses/longestValidParentheses.py
# Python program to find length of the longest valid substring
 
class Solution: 
 def longestValidParentheses(self,string):
     n = len(string)

     # Create a stack and push -1
     stack_ = []
     stack_.append(-1)

     # Initialize result
     result = 0

     # Traverse all characters of given string
     for i in range(n):

         # If "(" then,..push index of it
         if string[i] == '(':
             stack_.append(i)

         # If ")",then..
         else:   

             # Pop the previous opening bracket's index
             if len(stack_) != 0:
                stack_.pop()

             # Check if this length formed with base of
             # current valid substring is more than max 
             if len(stack_) != 0:
                 result = max(result, 
                              i - stack_[len(stack_)-1])

             # If stack is empty. push current index as
             # base for next valid substring (if any)
             else:
                 stack_.append(i)

     return result
